external_services/lti1p3.py

```diff
@@ -0,0 +1,350 @@
+import jwt
+import json
+import logging
+import datetime
+from typing import List, Dict, Any
+
+from django.conf import settings
+from django.contrib.auth.models import User
+from django.core.exceptions import PermissionDenied
+from django.urls import reverse
+from aplus.api import api_reverse
+
+from jwcrypto import jwk
+from oauthlib.openid import RequestValidator as OIDCRequestValidator
+from oauthlib.common import Request as OAuthRequest
+
+from lib.helpers import build_aplus_url
+from exercise.exercise_models import LTI1p3Exercise
+from course.models import CourseInstance
+from .models import LTI1p3Service
+from authorization.models import OAuth2BearerToken
+
+
+logger = logging.getLogger('aplus.external_services')
+
+def _add_claim(id_token: dict, key: str, value: Any) -> None:
+    id_token.update({f"https://purl.imsglobal.org/spec/lti/claim/{key}": value})
+
+
+class LTI1p3InitiateLogin:
```

Comment:
I don't think the class looks that bad. Even though only one method is currently used, classes combine the data together and it is possible to add more methods (if there were needs for that).

Reply:
Yep, it is quite simple, but follows the same approach as the old LTI, as called from views.py

Reply:
Here is a long text about class philosophy and how it applies to this particular class. It might read a bit harsh but don't mind that, it is just how text naturally comes to me. TLDR is that it isn't clear to me what the class represents. The name should't be a verb (e.g. `LTI1p3InitiateLoginRequest`, copying the old LTIRequest, would make more sense), and the doc string needs to change because it makes it seem like the class itself is handling sending the authentication request back. With those changes I'd find the class acceptable but I'd still rather it just be function: I'd find the code clearer if it just called a function (e.g. `get_lti1p3_parameters` or something similar) and it can be made into a class later if more data or methods need to be added.

> Even though only one method is currently used, classes combine the data together and it is possible to add more methods

In general I would agree with this but a class needs to be (1) a clear representation of some concept and (2) either encapsulate data related to the concept together (i.e. multiple attributes) or contains a set of methods that operates on the same data (the object's contents). Without those requirements, you would end up making every function into a class. Making a class because it might fulfill these conditions later doesn't make sense: just make the class later if needed instead of potentially having a class that doesn't fulfill the requirements. A rare exception for the first requirement to be enough is that if it makes the code easier to understand. The reason why this is rare is that without the second requirement, it is basically a function masquerading as a class which typically makes the code harder to understand.

In this case, the class only has a single useful attribute (I don't count self.service as that is not used anywhere) and a single method that just works as a getter for that attribute. This clearly fails the second requirement.

As for the exceptional case of making the code more understandable, I just find this class does the opposite. The name makes it sound like the class initiates a login but all it does is extract parameters into a list. It is also weird that a class name is a verb as it is the class' methods that are supposed to do things, not the class itself. At the very least the doc string should change. It makes it sound like the class itself handles sending the authentication request back. 

As for the old LTI (I presume you mean LTIRequest), it is a bit questionable but it does have a much clearer name and multiple methods that do things, so I can somewhat understand what it represents and why it exists.

Reply:
I don't have strong preferences, either. Made it a function, then.